omni-relayer/src/utils/fee.rs

```diff
@@ -32,49 +40,122 @@ pub async fn get_token_decimals(jsonrpc_client: &JsonRpcClient, token: &AccountI
-        .ok_or_else(|| anyhow::anyhow!("Failed to get price for address: {}", address))
+fn get_symbol_and_decimals_by_chain(chain: ChainKind) -> (&'static str, u32) {
+    match chain {
+        ChainKind::Eth => ("ethereum", 18),
+        ChainKind::Near => ("near", 24),
+        ChainKind::Sol => ("solana", 9),
+        ChainKind::Arb => ("arbitrum", 18),
+        ChainKind::Base => ("base", 18),
+    }
 }
 
-pub async fn is_fee_sufficient(jsonrpc_client: &JsonRpcClient, sender: &OmniAddress, recipient: &OmniAddress, token: &AccountId, fee: u128) -> Result<bool> {
-    let token_price = get_price_by_contract_address("near-protocol", token.as_ref()).await?;
-    let token_decimals = get_token_decimals(jsonrpc_client, token).await?;
-
-    let given_fee = fee as f64 / 10u128.pow(token_decimals) as f64 * token_price;
+async fn calculate_price(amount: u128, symbol: &str, decimals: u32) -> Result<f64> {
+    Ok(amount as f64 / 10u128.pow(decimals) as f64 * get_price_by_symbol(symbol).await?)
+}
 
-    // TODO: Right now I chose a random fee (around 0.10 USD), but it should be calculated based on the chain in the future
-    let sender_fee = match sender {
-        OmniAddress::Near(_) => 0.03 * get_price_by_symbol("near").await?,
-        OmniAddress::Eth(_) => 0.00005 * get_price_by_symbol("ethereum").await?,
-        OmniAddress::Sol(_) => 0.001 * get_price_by_symbol("solana").await?,
-        OmniAddress::Arb(_) | OmniAddress::Base(_) => todo!()
+pub async fn is_fee_sufficient(
+    config: &config::Config,
+    jsonrpc_client: &JsonRpcClient,
+    sender: &OmniAddress,
+    recipient: &OmniAddress,
+    token: &AccountId,
+    fee: &Fee,
+) -> Result<bool> {
+    let (sender_symbol, sender_token_decimals) =
+        get_symbol_and_decimals_by_chain(sender.get_chain());
+    let (recipient_symbol, recipient_token_decimals) =
+        get_symbol_and_decimals_by_chain(recipient.get_chain());
+
+    let native_fee_usd =
+        calculate_price(fee.native_fee.0, sender_symbol, sender_token_decimals).await?;
+
+    let fee_token_decimals = get_token_decimals(jsonrpc_client, token).await?;
+    let token_fee_usd = calculate_price(fee.fee.0, sender_symbol, fee_token_decimals).await?;
+
+    let expected_sender_fee_usd = match sender {
+        OmniAddress::Eth(_) => {
+            let http_provider = ProviderBuilder::new().on_http(
+                config
+                    .evm
+                    .rpc_http_url
+                    .parse()
+                    .context("Failed to parse ETH rpc provider as url")?,
+            );
+            calculate_price(
+                config.evm.init_transfer_gas_estimation * http_provider.get_gas_price().await?,
+                sender_symbol,
+                sender_token_decimals,
+            )
+            .await?
+        }
+        OmniAddress::Near(address) => {
+            if !storage::has_storage_deposit(jsonrpc_client, token, &address.parse::<AccountId>()?)
```

Comment:
@frolvanya If sender is sending the token is it possible for them not to have storage deposit? (Wouldn't it be required to hold the token in the first place?)

Reply:
Fixed it in https://github.com/Near-One/omni-bridge/pull/81/commits/0dec505603bc60e1b46585376af33f46d61b2d6c. In fact we can completely ignore all charges that come from transfer initialisation since user pays for it